language-support/ts/codegen/src/TsCodeGenMain.hs

```diff
@@ -47,63 +47,75 @@ optionsParserInfo = info (optionsParser <**> helper)
     <> progDesc "Generate TypeScript bindings from a DAR"
     )
 
+readPackages :: [FilePath] -> IO [(PackageId, (Package, Maybe String))]
+readPackages dars = concatMapM readPackage dars
+  where
+    readPackage dar = do
+      dar <- B.readFile dar
+      let archive = Zip.toArchive $ BSL.fromStrict dar
+      dalfs <- either fail pure $ DAR.readDalfs archive
+      DAR.DalfManifest{packageName} <- either fail pure $ DAR.readDalfManifest archive
+      forM ((DAR.mainDalf dalfs, packageName) : map (, Nothing) (DAR.dalfs dalfs)) $
+        \(dalf, mbPkgName) -> do
+          (pkgId, pkg) <- either (fail . show)  pure $ Archive.decodeArchive Archive.DecodeAsMain (BSL.toStrict dalf)
+          pure (pkgId, (pkg, mbPkgName))
+
+mergePackageMap :: [(PackageId, (Package, Maybe String))] -> Either String (Map.Map PackageId (Maybe String, Package))
+mergePackageMap ps = foldl' merge (Right Map.empty) ps
```

Comment:
I think this is still way more complex and harder to read than it needs to be, how about the following:

```haskell
mergePackageMap :: [(PackageId, (Package, Maybe String))] -> Either String (Map.Map PackageId (Maybe String, Package))
mergePackageMap ps = foldlM merge Map.empty ps
  where
    merge :: Map.Map PackageId (Maybe String, Package) -> (PackageId, (Package, Maybe String)) -> Either String (Map.Map PackageId (Maybe String, Package))
    merge pkgs (pkgId, (pkg, mbPkgName)) = do
        let pkgNames = mapMaybe fst (Map.elems pkgs)
        -- Check if there is a package with the same name but a different package id.
        whenJust mbPkgName $ \name -> when (pkgId `Map.notMember` pkgs && name `elem` pkgNames) $
            Left $ "Duplicate name '" <> name <> "' for different packages detected"
        let update mbOld = case mbOld of
                Nothing -> pure (Just (mbPkgName, pkg))
                Just (mbOldPkgName, _) -> do
                    -- Check if we have colliding names for the same package
                    whenJust (liftA2 (,) mbOldPkgName mbPkgName) $ \(name1, name2) ->
                        when (name1 /= name2) $ Left $ "Different names ('" <> name1 <> "' and '" <> name2 <> "') for the same package detected"
                    pure (Just (mbOldPkgName <|> mbPkgName, pkg))
        Map.alterF update pkgId pkgs
```

Reply:
Very nice. Virtuoso code.